crates/dbsp/src/circuit/dbsp_handle.rs

```diff
@@ -828,4 +961,412 @@ mod tests {
+
+    /// If we call commit, we should preserve the checkpoint list across circuit
+    /// restarts.
+    #[test]
+    fn checkpoint_file() {
+        let (_temp, cconf) = mkconfig();
+
+        {
+            let (mut dbsp, (_input_handle, _output_handle, sample_size_handle)) =
+                mkcircuit(&cconf).unwrap();
+            sample_size_handle.set_for_all(2);
+            dbsp.step().unwrap();
+            dbsp.commit_named("test-commit").expect("commit failed");
+            dbsp.step().unwrap();
+            dbsp.commit().expect("commit failed");
+        }
+
+        {
+            let (mut dbsp, _) = mkcircuit(&cconf).unwrap();
+            let cpm = &dbsp.checkpointer.list_checkpoints().unwrap()[0];
+            assert_eq!(cpm.step_id, 1);
+            assert_ne!(cpm.uuid, Uuid::nil());
+            assert_eq!(cpm.identifier, Some(String::from("test-commit")));
+
+            let cpm2 = &dbsp.checkpointer.list_checkpoints().unwrap()[1];
+            assert_eq!(cpm2.step_id, 2);
+            assert_ne!(cpm2.uuid, Uuid::nil());
+            assert_ne!(cpm2.uuid, cpm.uuid);
+            assert_eq!(cpm2.identifier, None);
+        }
+    }
+
+    /// We should fail if we instantiate a circuit with the same storage
+    /// directory twice.
+    #[test]
+    fn circuit_takes_ownership_of_storage_dir() {
+        let (_temp, cconf) = mkconfig();
+        let (_dbsp, _) = mkcircuit(&cconf).unwrap();
+
+        let r = Runtime::init_circuit(cconf, |_circuit| Ok(()));
+        assert!(matches!(
+            r,
+            Err(DBSPError::Storage(StorageError::StorageLocked(_, _)))
+        ));
+    }
+
+    /// We should fail if we revert to a checkpoint that doesn't exist.
+    #[test]
+    fn revert_to_unknown_checkpoint() {
+        let (_temp, mut cconf) = mkconfig();
+        let (dbsp, _) = mkcircuit(&cconf).unwrap();
+        drop(dbsp); // makes sure we can take ownership of storage dir again
+
+        cconf.init_checkpoint = Uuid::now_v7(); // this checkpoint doesn't exist
+
+        let res = mkcircuit(&cconf);
+        assert!(matches!(
+            res,
+            Err(DBSPError::Storage(StorageError::CheckpointNotFound(_)))
+        ));
+    }
+
+    /// We panic if we initialize to a partially complete checkpoint.
+    #[test]
+    #[should_panic]
+    fn revert_to_partial_checkpoint() {
+        let (temp, mut cconf) = mkconfig();
+        let (dbsp, _) = mkcircuit(&cconf).unwrap();
+        drop(dbsp); // makes sure we can take ownership of storage dir again
+
+        cconf.init_checkpoint = Uuid::now_v7(); // A made-up checkpoint, that does not have the necessary files
+        let checkpoint_dir = temp.path().join(cconf.init_checkpoint.to_string());
+        create_dir_all(checkpoint_dir).expect("can't create checkpoint dir");
+
+        // Initializing this circuit again will panic because it won't find the
+        // necessary files in the checkpoint directory.
+        let (_dbsp, _) = mkcircuit(&cconf).unwrap();
```

Comment:
how do you know the panic happens here and not earlier?
maybe you can check the panic message?

Reply:
The failure will happen in any worker thread that discovers things aren't complete and panic. We do get the error back through the panic handler of the runtime. But we don't currently have the machinery to propagate errors back in this case to the handle/main thread. Maybe I will need to add it in the future.